### ISSUE TITLE
Autotracking with Axis Camera

### ISSUE THREAD

**Issue by alexyao2015** [blakeblackshear/frigate#10184]
**Background**
I am attempting to use autotracking with an Axis ptz camera (AXIS M5525-E), but it started spewing out errors even though the camera has all the onvif capabilities and is supported. Since [Axis was one of the founders of the onvif standard](https://www.axis.com/onvif), I was pretty confident that their implementation should have been correct. 

**Debugging**
Running the test script showed 3 YES logs, however the service_capabilities was empty. When running Frigate, it [crashed at this line](https://github.com/blakeblackshear/frigate/blob/5028a9632e03262c8adb5a4e7afdf5f995b47f87/frigate/ptz/onvif.py#L545) where obtained the service_capabilities and received a None value. I removed this check and was able to get past it.

Next I noticed my logs giving some error relating to the [zoom being out of bounds at this line](https://github.com/blakeblackshear/frigate/blob/5028a9632e03262c8adb5a4e7afdf5f995b47f87/frigate/ptz/onvif.py#L638). I determined that my camera responded with 0-1 as the values Frigate checks for zoom interpolation, however my camera supported from 0-0.499. I was able to fix this using the following:

```
zoom = numpy.interp(
    zoom,
    [0, 1],
    [
        self.cams[camera_name]["zoom_limits"]["Range"]["XRange"]["Min"],
        self.cams[camera_name]["zoom_limits"]["Range"]["XRange"]["Max"],
    ],
)
```

After these two changes, autotracking started working normally.

**Fix**

1. We may want to use a different check instead of service_capabilities to verify movement status. Why wasn't `GetStatus` simply used and a check to see if that was idle? Is there a reason why a completely different value is checked?
2. For the camera interpolation, I would suggest using the above lines of code to determine the zoom range to interpolate. I only have 1 camera, so I cannot verify that this works across other cameras and would need others to help verify if this is a good solution.

CC: @hawkeye217

**Comment by hawkeye217** [comment]
Thanks for the info.

For #1: There was a [recent discussion](https://github.com/blakeblackshear/frigate/discussions/9927#discussioncomment-8525067) about `GetServiceCapabilities` and AXIS cameras.

The ONVIF protocol says [here](https://www.onvif.org/specs/srv/ptz/ONVIF-PTZ-Service-Spec-v1712.pdf#%5B%7B%22num%22%3A97%2C%22gen%22%3A0%7D%2C%7B%22name%22%3A%22XYZ%22%7D%2C69%2C728%2C0%5D) that the `GetServiceCapabilities` call should return True if the camera supports `MoveStatus` information. This `MoveStatus` parameter is used by Frigate's autotracking to determine when your PTZ is moving.

The reason why this parameter was chosen was:
1. It is following the ONVIF standard.
2. Some cameras report bogus values for `GetStatus` - they may show `MOVING` or `IDLE` but because `GetServiceCapabilities` doesn't indicate support for `MoveStatus`, the values from `GetStatus` are untrustworthy.

Other cameras we've tested with and that users have reported success with all have correctly reported `GetServiceCapabilities` as expected by the ONVIF standard.

For #2: the ONVIF code in Frigate saves the absolute zoom limits from the camera [here](https://github.com/blakeblackshear/frigate/blob/5028a9632e03262c8adb5a4e7afdf5f995b47f87/frigate/ptz/onvif.py#L262) and then uses them in calculations elsewhere when absolute zooming is used. In my testing, the `zoom_limits` values are not always the same as the absolute zoom limit values reported in the ptz profile's `Spaces.AbsoluteZoomPositionSpace` object. 

In short, I've found that the ONVIF "standard" is not quite as standard as one would hope. Basic things (continuous move, presets, streams, etc) seem to be widely standardized, but the implementation of some of the finer parameters seem to vary widely from manufacturer to manufacturer, despite them saying they "support ONVIF".

**Comment by alexyao2015** [comment]
Thanks for your comments! I understand the concerns now with some cameras not conforming to the ONVIF spec. I still do find it odd that a manufacturer like AXIS produces cameras that doesn't confine to a spec they helped create. Ultimately, I as well as others likely would want their cameras to work with this new feature, so I would like to toss out a couple of ideas for solving these issues.

For `GetServiceCapabilities`, could we leave this as an optional config options to simply ignore this check? I did see that in my specific case, disabling this check allowed the camera to still function as `GetStatus` did return the correct `MoveStatus` values. 

As for the zoom limits, I know you mention that `zoom_limits` are different, but does this cause issues with other cameras if it were used instead of the `Spaces.AbsoluteZoomPositionSpace` object? I know it certainly worked on my camera, but I cannot speak to other cameras.

Additionally as I was taking look as this, I think I found a bug. The usage of `np.interp` is [inverted here](https://github.com/blakeblackshear/frigate/blob/dev/frigate/ptz/onvif.py#L633-L641). Since the values are getting interpolated back from the camera's view to 0-1, the camera's zoom should come first.

Hopefully we can figure out a solution to getting something working in for everyone. I have my frigate version monkey patched with the few changes I mentioned and they have been working well on my camera.

**Comment by hawkeye217** [comment]
> Thanks for your comments! I understand the concerns now with some cameras not conforming to the ONVIF spec. I still do find it odd that a manufacturer like AXIS produces cameras that doesn't confine to a spec they helped create. Ultimately, I as well as others likely would want their cameras to work with this new feature, so I would like to toss out a couple of ideas for solving these issues.
> 
> For `GetServiceCapabilities`, could we leave this as an optional config options to simply ignore this check? I did see that in my specific case, disabling this check allowed the camera to still function as `GetStatus` did return the correct `MoveStatus` values.
> 

Adding features to work around individual camera issues or manufacturer's firmware bugs and incomplete ONVIF support isn't something that makes a lot of sense from a support perspective, especially when myself and the other maintainers don't have those cameras to test with. 

> As for the zoom limits, I know you mention that `zoom_limits` are different, but does this cause issues with other cameras if it were used instead of the `Spaces.AbsoluteZoomPositionSpace` object? I know it certainly worked on my camera, but I cannot speak to other cameras.
> 
> Additionally as I was taking look as this, I think I found a bug. The usage of `np.interp` is [inverted here](https://github.com/blakeblackshear/frigate/blob/dev/frigate/ptz/onvif.py#L633-L641). Since the values are getting interpolated back from the camera's view to 0-1, the camera's zoom should come first.
> 

This is incorrect and not a bug. Absolute zoom values from the camera get mapped to a value between 0 and 1. Frigate uses a value of 0-1 internally to keep track of cameras using absolute zooming, so the code as written is working as intended. 

However, it may be possible to better catch errors or unexpected values with cameras like yours. Could you enable debug logging, restart Frigate, and paste the initial log output as well as your Frigate config here?

```
logger:
  default: info
  logs:
    frigate.ptz.autotrack: debug
    frigate.ptz.onvif: debug
```

**Comment by alexyao2015** [comment]
@hawkeye217 

Sorry about the long delay here. I wanted to give an update on this as I have been patching frigate manually each release and have been using autotracking on these cameras without issue since I made the initial post. I am confident that these cameras can be supported, even if they do not support the "correct" ONVIF standard. With these patches applied, autotracking is perfect and I've observed it track people for months without any fail. 

Attached in the zip file, I've included the relevant log snippets and config snippets as well as the git patches to `onvif.py` and `autotrack.py` that I've been using. I would imagine that the onvif profile from the logs and the patches will likely prove to be the most useful. The config is pretty standard and functions just fine with my camera, post patch.

I'm hoping that these files will allow something to make it into the main branch, but they will not suffice as I have them written today because they assume some specifics about my particular cameras. 

Thanks for all your work on this!
[data.zip](https://github.com/user-attachments/files/17238271/data.zip)

**Comment by alexyao2015** [comment]
Some answers to your response:

> especially when myself and the other maintainers don't have those cameras to test with

I am willing to help with the code and test this if necessary if there was a selection that allows the selection of this `zoom_limits` in the ONVIF properties compared to the `absolute_zoom_range` that is used today. I am very confident that this is all that is required to support at least one other camera type. 

>  Absolute zoom values from the camera get mapped to a value between 0 and 1. Frigate uses a value of 0-1 internally to keep track of cameras using absolute zooming, so the code as written is working as intended.

Here is a simple demonstration showing that the code is wrong. If this is incorrect, please let me know which of my assumptions are incorrect. Additionally, as mentioned I've been running a patched version with this exact code and haven't had any adverse side effects whereas it was completely nonfunctional before.

```
import numpy as np
# In this example, our zoom level is 0.25 within a camera zoom range of 0-0.5 where 0 is 0% zoom and 0.5 is 100% zoom
# 0.25 is 50% of our zoom so we expect the result of this operation to return 0.5.

# Before (Code in repo today)
# Here the result of this operation is 0.25 because we are scaling to the camera zoom level rather than frigate's zoom level
np.interp(
  0.5, # our zoom level to interpolate
  [
    0, # frigate min zoom -> always 0
    1, # frigate max zoom -> always 1
  ],
  [
    0, # camera min zoom -> read onvif from camera
    0.5, # camera max zoom -> read onvif from camera
  ]
)
# After (my changes)
# Here the result of the operation is 0.5 as we expect
np.interp(
  0.25, # our zoom level to interpolate
  [
    0, # camera min zoom -> read onvif from camera
    0.5, # camera max zoom -> read onvif from camera
  ],
  [
    0, # frigate min zoom -> always 0
    1, # frigate max zoom -> always 1
  ]
)
```

**Comment by hawkeye217** [comment]
Thanks for this! A few thoughts:

> I am willing to help with the code and test this if necessary if there was a selection that allows the selection of this zoom_limits in the ONVIF properties compared to the absolute_zoom_range that is used today. I am very confident that this is all that is required to support at least one other camera type.

The `absolute_zoom_range` is a property that I assign in code directly from the onvif property `AbsoluteZoomPositionSpace`, per the onvif spec. See the code here: https://github.com/blakeblackshear/frigate/blob/dev/frigate/ptz/onvif.py#L297

Some cameras have the ability to configure zooming limits in firmware, so the `ZoomLimits` onvif property reflects this. Again, from the onvif spec:

> ZoomLimits – The zoom limits element should be present for a PTZ node that supports absolute zoom. If the element is present it signals the supports for configurable zoom limits. If limits are enabled the zoom movements shall always stay within the specified range

On my Dahua, the `AbsoluteZoomPositionSpace` property works as I'd expect and mirrors `ZoomLimits`. However on other cameras I've seen, `ZoomLimits` is not always present (again, because of a bad onvif implementation), hence why the code uses `AbsoluteZoomPositionSpace` to determine the usable range for absolute zooming. 

Your cameras do report values for `AbsoluteZoomPositionSpace`, so you shouldn't need to patch that code. But because your camera's onvif implementation doesn't support `GetServiceCapabilities`, you'll need to continue to patch for that as Frigate will continue to follow the onvif standard there.

As for the interpolation logic, it does seem to be reversed, thanks for catching that. I'll have a PR up to fix that shortly.

**Comment by hawkeye217** [comment]
Once again I just have to say that the onvif standard and the implementation of it from camera manufacturers is a mess. I did a little more reading and I can see how `ZoomLimits` relates to `AbsoluteZoomPositionSpace`, but the fact that it's up to the camera manufacturer to determine how these two values should practically correlate leaves developers who want to use the standard across various cameras hanging out to dry.

The general direction Frigate has taken with the codebase is to avoid hardware or camera-specific code to work around bugs or problems. Since you're already familiar with patching, that's our recommendation to keep your cameras working. I don't suspect there will be any major changes to the autotracking or onvif code, so you shouldn't have a difficult time patching from release to release.

**Comment by putteg** [comment]
> [@hawkeye217](https://github.com/hawkeye217)
> 
> Sorry about the long delay here. I wanted to give an update on this as I have been patching frigate manually each release and have been using autotracking on these cameras without issue since I made the initial post. I am confident that these cameras can be supported, even if they do not support the "correct" ONVIF standard. With these patches applied, autotracking is perfect and I've observed it track people for months without any fail.
> 
> Attached in the zip file, I've included the relevant log snippets and config snippets as well as the git patches to `onvif.py` and `autotrack.py` that I've been using. I would imagine that the onvif profile from the logs and the patches will likely prove to be the most useful. The config is pretty standard and functions just fine with my camera, post patch.
> 
> I'm hoping that these files will allow something to make it into the main branch, but they will not suffice as I have them written today because they assume some specifics about my particular cameras.
> 
> Thanks for all your work on this! [data.zip](https://github.com/user-attachments/files/17238271/data.zip)

Hey @alexyao2015 ,
Can you please tell me how to patch Frigate with your patch?

@hawkeye217  Please let me know if you need a Axis camera to test on. Maybe I can help you with that.

**Comment by lajmeen** [comment]
@alexyao2015 Would you please explain (or point me in the right direction) how to patch this Axis PTZ-fix? I have the same issue that Frigate-crashes when trying to enable auto-tracking in the ONVIF-config.
I've managed to fork the repo and made the changes you suggested but nothing appears when I add the forked repo under custom addons.
Much appreciated!

**Comment by ehn** [comment]
> Once again I just have to say that the onvif standard and the implementation of it from camera manufacturers is a mess. I did a little more reading and I can see how `ZoomLimits` relates to `AbsoluteZoomPositionSpace`, but the fact that it's up to the camera manufacturer to determine how these two values should practically correlate leaves developers who want to use the standard across various cameras hanging out to dry.
> 
> The general direction Frigate has taken with the codebase is to avoid hardware or camera-specific code to work around bugs or problems. Since you're already familiar with patching, that's our recommendation to keep your cameras working. I don't suspect there will be any major changes to the autotracking or onvif code, so you shouldn't have a difficult time patching from release to release.

On the one hand, I completely understand that Frigate developers don't want to add and maintain hacks for all the various broken ONVIF implementations, all broken in their own way. On the other hand, this currently limits the cameras that can be used with autotracking to a very small selection.

I have a bunch of Tapo C200 and C210 cameras. They're cheap and crappy. I was not surprised to find out that autotracking doesn't work for them. I also have a couple of Hanwha XNP-6550RH cameras. They're in the $4,000 segment and generally excellent. Alas, their ONVIF implementation is also not up to par, and I was quite disappointed to learn that Frigate's autotracking won't work for them either.

In this thread, we have learned that the same is true for Axis cameras, despite the company being part of the group that created the ONVIF standard.

Would it make sense to create a separate project that provides an ONVIF proxy? Then it could implement all the hacks needed for various cameras, including ones that don't support ONVIF at all but have other protocols or APIs for PTZ control, while providing a clean, ONVIF-compliant interface for Frigate (and other consumers). It could also provide a non-ONVIF API for common camera features not covered by ONVIF. Essentially, it could do for camera control what go2rtc does for video streams.

**Comment by hawkeye217** [comment]
> Would it make sense to create a separate project that provides an ONVIF proxy?

That's a great idea. I don't have the time or the hardware to implement and maintain anything like this, but I'd be happy to consult with anyone who does to provide best practices for using something like it with Frigate. 

To anyone who is interested, feel free to open a separate discussion to continue the dialogue.

**Comment by markdmann** [comment]
> [@hawkeye217](https://github.com/hawkeye217)
> 
> Sorry about the long delay here. I wanted to give an update on this as I have been patching frigate manually each release and have been using autotracking on these cameras without issue since I made the initial post. I am confident that these cameras can be supported, even if they do not support the "correct" ONVIF standard. With these patches applied, autotracking is perfect and I've observed it track people for months without any fail.
> 
> Attached in the zip file, I've included the relevant log snippets and config snippets as well as the git patches to `onvif.py` and `autotrack.py` that I've been using. I would imagine that the onvif profile from the logs and the patches will likely prove to be the most useful. The config is pretty standard and functions just fine with my camera, post patch.
> 
> I'm hoping that these files will allow something to make it into the main branch, but they will not suffice as I have them written today because they assume some specifics about my particular cameras.
> 
> Thanks for all your work on this! [data.zip](https://github.com/user-attachments/files/17238271/data.zip)

@alexyao2015 Hey, can you give a quick blurb on how you are patching the files?  I need to do this for my axis cameras.  Thanks!

**Comment by hawkeye217** [comment]
> > [@hawkeye217](https://github.com/hawkeye217)
> > Sorry about the long delay here. I wanted to give an update on this as I have been patching frigate manually each release and have been using autotracking on these cameras without issue since I made the initial post. I am confident that these cameras can be supported, even if they do not support the "correct" ONVIF standard. With these patches applied, autotracking is perfect and I've observed it track people for months without any fail.
> > Attached in the zip file, I've included the relevant log snippets and config snippets as well as the git patches to `onvif.py` and `autotrack.py` that I've been using. I would imagine that the onvif profile from the logs and the patches will likely prove to be the most useful. The config is pretty standard and functions just fine with my camera, post patch.
> > I'm hoping that these files will allow something to make it into the main branch, but they will not suffice as I have them written today because they assume some specifics about my particular cameras.
> > Thanks for all your work on this! [data.zip](https://github.com/user-attachments/files/17238271/data.zip)
> 
> [@alexyao2015](https://github.com/alexyao2015) Hey, can you give a quick blurb on how you are patching the files? I need to do this for my axis cameras. Thanks!

So the issue we've been [discussing over on Reddit](https://www.reddit.com/r/frigate_nvr/comments/1gkknku/comment/mnzu6nz/) is actually not the same issue as this thread. Your issue isn't something Frigate can correct or work around.
